### DISCUSSION TITLE
Next.js Static Site Generation (SSG) question

### DISCUSSION THREAD

**Discussion by Dev4LifeV** [vercel/next.js#66100] [Help] [answered]
### Summary

I've been discussing the Next.js SSG generation with my team, and I'm quite confused.

I haven't found this information in the web out there.

I've been doing some researches through AI and I got this response:

"Yes, that’s correct. When you generate a static site using Next.js (or similar frameworks), the site is initially served as static HTML to the browser.

However, after the static content is served, Next.js “rehydrates” the static HTML into a fully interactive React application on the client side. This process is often referred to as hydration."

So, whenever I have dynamic components (for example, I have a component that applies a style when hovered, so I use a state to define a boolean to check whether it is hovered or not), will React be managing these components through these listeners?

Is this correct?

Next.js => HTML => React (on needed components)

I'm asking this because we are discussing whether or not we should pass an object as a prop in a component instead of inline props to improve performance.

I'm trying to ensure that the result of a Next.js Static Site in the end is whether or not a static HTML with some React listeners around it.

**Comment by icyJoseph** [comment]
I think you can simplify the mental model here. Next.js has a mode where it can, at build time, pre-render a page into an HTML file, with JavaScript, CSS assets, and JSON (let's keep it only to those).

This means that, any entity that requests your page, will see as first frame, an already pre-rendered page. This has some implications, for example, bots and SEO tooling can more easily digest your content. People who have JavaScript disabled, can also, to some extend, consume your application.

However, React is still shipped, as said above, JavaScript is still part of the build output. 

React at the beginning, does either, a render pass where it calls your React component tree, and mounts the results into the DOM, or it calls your React component tree, and walks an already existing DOM (from the build time pre-render), matching it to what it gets from rendering your component tree, and it attaches event listeners, wherever the React component tree says so. This is more or less, called, Hydration. Bringing life to an otherwise "dry" piece of HTML. After that, React goes idle.

Next.js is setup to do the hydration step.  So if you have a page that renders a counter with initial state zero. You'll have a HTML document, where you have the HTML tags, and buttons that make up the counter. And after hydration is done, you'll be able to click the buttons to update the counter in the UI. 

Now imagine, that the initial state is actually calculated from some database query, at a parent of your counter, and passed as props, then at build time, the query will be made, given that the SSG parts are in place, and your counter will start not at zero, but at the initial state queried at build time. 

The interesting thing here is that, the data from the parent of the counter, (or result of GetStaticProps), are collected as JSON, and are included in the page's initial HTML, either as RSC or as __NEXT_DATA__ JSON script tag, and these are used by the hydration step, to calculate the initial props that go into the counter.

Another implication, is that you can have parts of the React tree be deferred, from the build time rendering step, and only render once in the client. After hydration, rather than having React going idle, you can active a flag within a `useEffect`, that would in turn unlocking rendering of even more components. See here, https://demystifying-rsc.vercel.app/client-components/no-ssr/

I must say though, your question is not entirely clear. Are you concerned, or not, that React still does do something to your app, even when pre-rendered?

**Reply by Dev4LifeV** [reply]
Sorry for the lack of clarity.

I'm concerned because, since I'm working with SSG, I don't know whether it makes sense to be concerned about React's performance.

Let's assume the example below: I have a parent component with a child component. This child component is receiving an object props:

```JSX

<ParentComponent> //has some state
    <ChildComponent props={{hello: "there"}}/>
</ParentComponent>

```

Whenever the ParentComponent is updated, it will be re-rendered, then ChildComponent will also be re-rendered so the object will be recreated. To avoid this, a useMemo should be used to wrap this object.

(Please correct me if I'm wrong)

So, instead of this object, we could use inline props to avoid the need for the hook and all this stuff.

Since I'm using SSG, should I be concerned about this? Should I consider React's performance?

**Reply by icyJoseph** [reply, marked as answer]
React's very fast. It is essentially a while loop just running through your components.

The standard React rendering behavior is to render the sub tree, from which state was changed, or if useSyncExternalStore is invoked. From that point downwards, React collects plain object representations of the UI your components want to show in the screen.

- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#standard-render-behavior

Once it has this representation, it diffs with the UI, and if nothing has changed, nothing is committed to the DOM, and nothing happens. In the pixel pipeline, https://web.dev/articles/rendering-performance, in the standard case, we stop at the JavaScript step. That is to say, that React rendering is not the same as Browser Paint. Otherwise, it makes the necessary changes to the DOM to synchronize.

When does the slowness creep in? When the rendering work is itself too heavy. Developers often corner themselves into it by excessively setting state, through useEffect, and other means, that waterfall even more rendering runs. However, it is also possible to run into this, when resolving asynchronous data, be it network requests or event driven data, for example, typically, you have `isLoading`, going from true to false, and in the middle, other pieces of state might get updated, which in turn causes a bunch of renders to happen.

To control these situations, we have `useMemo`, `useCallback`, and `memo`, which are ways to prune the rendering work done by React, by telling it that, certain subtrees, or effects, can be skipped, because we haven't changed the state/references that are passed to it. 

- https://react.dev/learn/react-compiler the React compiler will make these functions unnecessary 😄 

Now, React is also very clever when it comes to rendering and composition. In your example, if `ParentComponent` renders, `ChildComponent` doesn't, necessarily have to render. Say:

```jsx
const ParentComponent = ({ children }) => {
    const [state, setState] = useState(initial);
   
    return (
       <div>
          <CalendarComponent state={state} />
          <Controls setState={setState} />
          {children}
       </div>
    )
}
```

When the setState is called, only `CalendarComponent`, and `Controls` are re-rendered, in the standard behavior. If `Controls` is wrapped with memo, it will skip render too. And if `setState` is called with the same state, even `CalendarComponent` could skip rendering. 

What is definitely true, is that `children` doesn't render, in this case, ever, when `setState` is called. Sure, one could say that `CalendarComponent` or `Controls` are using a context provider, and themselves change that, and that in turn maybe affects `ClientComponent`, but that's a stretch. I am just mentioning it for completeness sake.

```jsx
const Home = (props) => {
    return (
      <ParentComponent>
          <ChildComponent {...props} />
      </ParentComponent>
    )
}
```

So what I mean to say, in the above, when parent component renders, child component can skip, because it is not Home that's rendering. 

You can think of it this way. ChildComponent is a hierarchical child of ParentComponent, but it is not owned by it. Both ParentComponent, and ChildComponent, are owned by Home, and render when Home renders, in the standard behavior. 

Whats' even cooler, is that ChildComponent, could even be a Server Component, that doesn't ever render in the client, React just knows how to weave/stitch the Server and Client tree together.

**Reply by Dev4LifeV** [reply]
Thank you very much for all the clarification, @icyJoseph! I will mark the discussion as answered. I really appreciated all the help.
